From: Michal Kubecek <mkubecek@suse.cz>
Date: Wed, 30 May 2018 11:16:15 +0200
Subject: ipv6: omit traffic class when calculating flow hash
Patch-mainline: v4.18-rc1
Git-commit: fa1be7e01ea863e911349e30456706749518eeab
References: bsc#1095042

Currently the flow hash is calculated from flowlabel member of struct
flowi6 which encodes both flow label and traffic class fields of IPv6
header. As traffic class can change within a TCP connection (which ssh
reportedly does), this can cause e.g. ECMP path switch.

As traffic class is not guaranteed to be constant within a flow, mask it
out when using flowi6::flowlabel for flow hash key and use only the actual
flow label.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 net/core/flow_dissector.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/net/core/flow_dissector.c b/net/core/flow_dissector.c
index 91c85422429c..ab96cfecb51e 100644
--- a/net/core/flow_dissector.c
+++ b/net/core/flow_dissector.c
@@ -735,7 +735,8 @@ __u32 __skb_get_hash_flowi6(struct sk_buff *skb, const struct flowi6 *fl6)
 	keys.ports.src = fl6->fl6_sport;
 	keys.ports.dst = fl6->fl6_dport;
 	keys.keyid.keyid = fl6->fl6_gre_key;
-	keys.tags.flow_label = (__force u32)fl6->flowlabel;
+	keys.tags.flow_label = (__force u32)(fl6->flowlabel &
+					     IPV6_FLOWLABEL_MASK);
 	keys.basic.ip_proto = fl6->flowi6_proto;
 
 	__skb_set_sw_hash(skb, flow_hash_from_keys(&keys),
@@ -843,7 +844,8 @@ __u32 __get_hash_from_flowi6(const struct flowi6 *fl6, struct flow_keys *keys)
 	keys->ports.src = fl6->fl6_sport;
 	keys->ports.dst = fl6->fl6_dport;
 	keys->keyid.keyid = fl6->fl6_gre_key;
-	keys->tags.flow_label = (__force u32)fl6->flowlabel;
+	keys->tags.flow_label = (__force u32)(fl6->flowlabel &
+					      IPV6_FLOWLABEL_MASK);
 	keys->basic.ip_proto = fl6->flowi6_proto;
 
 	return flow_hash_from_keys(keys);
-- 
2.17.1

